### PR TITLE
feat: upvote logic

### PR DESCRIPTION
**Pull request by JeraldYik**
## Context

Add upvote/undo upvote logic. Composite unique constraint added to foreign keys `postId` and `userId`.

Routes:
- `POST /upvote` (upvote)
- `DELETE /upvote` (undo upvote)

## Tests

- [x] Upvote a post
- [x] Try following the same post again, should throw a 400 error
- [x] Call `GET /posts` and `GET /posts/:id`. Check the new `upvoteCount: 1` attribute 
- [x] Undo upvote this post  
- [x] Call `GET /posts` and `GET /posts/:id`. Check `upvoteCount: 0`

### CODE REVIEW

File: backend/src/posts/posts.service.ts [hunk 1]
```diff
@@ -50,11 +62,14 @@ export class PostsService {
       commentsCount: model.comments.length,
       followsCount: model.follows.length,
       isFollowing: model.follows.some((follow) => follow.userId === user.id),
+      upvoteCount: model.upvotes.filter((upvote) => upvote.postId === model.id)
```
Comment (JeraldYik) [inline backend/src/posts/posts.service.ts:65]: just want to put it out here. is filtering necessary for tables joined using eager loading? should i just do a simple `model.upvotes.length` ?

File: backend/src/posts/posts.service.ts [hunk 2]
```diff
@@ -131,6 +151,8 @@ export class PostsService {
         canManage: user.id === post.user.id,
         isFollowing: post.follows.some((follow) => follow.userId === user.id),
         followsCount: post.follows.length,
+        upvoteCount: post.upvotes.filter((upvote) => upvote.postId === post.id)
```
Comment (JeraldYik) [inline backend/src/posts/posts.service.ts:154]: similar to above comment

**Review by JasonChong96 — CHANGES_REQUESTED** [review]

File: backend/src/posts/posts.service.ts [hunk 3]
```diff
@@ -77,6 +92,7 @@ export class PostsService {
         canManage: user.id === post.user.id,
         isFollowing: post.isFollowing,
         followsCount: post.followsCount,
+        upvoteCount: post.upvoteCount,
```
Comment (JasonChong96) [inline backend/src/posts/posts.service.ts:95]: Also we need an `isUpvoted` (or a better name) similar to `isFollowing`. This is to allow the frontend to know if the current user has already upvoted this post

**Review by JasonChong96 — CHANGES_REQUESTED** [review]

File: backend/src/posts/posts.service.ts (hunk 3 above)
Comment (JeraldYik) [reply backend/src/posts/posts.service.ts:100]: [commit](https://github.com/opengovsg/spotlightsg/pull/99/commits/edd74f4f2007d673fffa3befbe83b8e49360b78e). named it `hasBeenUpvoted`

File: backend/src/posts/posts.service.ts (hunk 1 above)
Comment (JeraldYik) [reply backend/src/posts/posts.service.ts]: [commit](https://github.com/opengovsg/spotlightsg/pull/99/commits/edd74f4f2007d673fffa3befbe83b8e49360b78e)

File: backend/src/posts/posts.service.ts (hunk 2 above)
Comment (JeraldYik) [reply backend/src/posts/posts.service.ts]: resolved with comment above

**Review by JasonChong96 — APPROVED** [review]
lgtm

**Pull request merged**